### PR TITLE
sort interviews by sources then names

### LINKED ISSUE #383: ”学生面试“页根据“来源”=>"姓名"排序

### PR DESCRIPTION
**Pull request by ada-only-one-ada**
Fixed #383. 
Firstly sort by sources. If no source, sort by name.
Screenshot 2024-10-03 at 12 32 53

### CODE REVIEW

File: src/pages/mentees.tsx [hunk 1]
```diff
@@ -111,8 +111,8 @@ function MenteeTable({ users, refetch }: {
         const yearB = menteeToYear.get(b.id) || "";
 
         if (yearA === yearB) {
-          const nameA = formatUserName(a.name, 'formal');
-          const nameB = formatUserName(b.name, 'formal');
+          const nameA = toPinyin(formatUserName(a.name, 'formal'));
```
Comment (weihanwang) [inline src/pages/mentees.tsx:114]: why need to change this?

File: src/pages/interviews.tsx [hunk 2]
```diff
@@ -145,18 +163,23 @@ function Applicants({ type, applicants, interviews, refetchInterviews,
...
 function Applicant({ type, applicant, interviews, refetchInterviews,
-  refetchApplicants
+  refetchApplicants, updateSource
  } : {
   type: InterviewType,
   applicant: User,
   interviews: Interview[],
   refetchInterviews: () => void,
   refetchApplicants: () => void,
+  updateSource:(id: string, source: string) => void,
```
Comment (weihanwang) [inline src/pages/interviews.tsx:173]: this function should not need parameter `id`

File: src/pages/interviews.tsx [hunk 3]
```diff
@@ -129,10 +133,24 @@ function Applicants({ type, applicants, interviews, refetchInterviews,
         </Tr>
       </Thead>
       <Tbody>
-        {applicants.map(a => 
+        {applicants
+          .sort((a1, a2) => {
```
Comment (weihanwang) [inline src/pages/interviews.tsx:137]: why not simply:

```
const comp = (sources[a1.id] || "").localeCompare(sources[a2.id]);
return comp !== 0 ? comp : (a1.name || "").localeCompare(a2.name);
```

File: src/pages/mentees.tsx [hunk 4]
```diff
@@ -111,8 +111,8 @@ function MenteeTable({ users, refetch }: {
         const yearB = menteeToYear.get(b.id) || "";
```
Comment (weihanwang) [inline src/pages/mentees.tsx:111]: why not simply:

```
const comp = (menteeToYear.get([a.id]) || "").localeCompare(menteeToYear.get([b.id]));
return comp !== 0 ? comp : (a.name || "").localeCompare(b.name);
```

File: src/pages/mentees.tsx (hunk 1 above)
Comment (weihanwang) [reply src/pages/mentees.tsx:114]: I see. In this case, add a function compareChinese in strings.ts similar to compareDate. Also add this link to the comment of this function to justify its existence.

Reuse this compareChinese in the `Applicants` function

File: src/pages/interviews.tsx (hunk 2 above)
Comment (ada-only-one-ada) [reply src/pages/interviews.tsx:173]: How do you get the source of each applicant to sort?` const comp = (sources[a1.id] || "").localeCompare(sources[a2.id]);`
You don't have source attribute on applicant.
Do you need to call  trpcNext.users.getApplicant.useQuery to get the source?
```
const source = (data?.application as Record<string, any> | null)
    ?.[menteeSourceField];
```

Comment (weihanwang) [reply src/pages/interviews.tsx:173]: The caller already knows the id of the applicant. At line 153, you simply do `updateSource={source => updateSource(a.id, source)}`

File: src/pages/interviews.tsx [hunk 5]
```diff
@@ -114,6 +114,10 @@ function Applicants({ type, applicants, interviews, refetchInterviews,
   refetchInterviews: () => void,
   refetchApplicants: () => void,
 }) {
+  const [sources, setSources] = useState<{ [id: string]: string }>({});
```
Comment (weihanwang) [inline src/pages/interviews.tsx:117]: add a comment "a map from applicants' user ids to sources"

File: src/pages/interviews.tsx [hunk 6]
```diff
@@ -129,10 +133,21 @@ function Applicants({ type, applicants, interviews, refetchInterviews,
         </Tr>
       </Thead>
       <Tbody>
-        {applicants.map(a => 
+        {applicants
+          .sort((a1, a2) => {
+            const comp = (!sources[a1.id] && sources[a2.id]) ? 1 :
+              (sources[a1.id] && !sources[a2.id]) ? -1 :
```
Comment (weihanwang) [inline src/pages/interviews.tsx:139]: these two lines are not needed. Simply `const comp = toPinyin(...).localeCompare(toPinyin(...))` suffices

File: src/pages/mentees.tsx [hunk 7]
```diff
@@ -106,17 +106,16 @@ function MenteeTable({ users, refetch }: {
       </Tr>
     </Thead>
     <Tbody>
+      {/* Need to convert it to pinyin, otherwise the result 
+      will not be correct if compare Chinese directly. Ref:
+      https://www.leevii.com/2023/04/about-the-inaccurate-chinese-sorting-of-localecompare.html */}
```
Comment (weihanwang) [inline src/pages/mentees.tsx:111]: move these comments to right above line 117

**Review by weihanwang — APPROVED** [review]

**Pull request merged**